changes to allow non-primary unique key associations in HR

gavinking opened the pull request:
see https://github.com/hibernate/hibernate-reactive/issues/565

This is a draft, which I need to think more about. But here so at least @Sanne and @DavideD can see what I'm thinking about doing.

gavinking commented on hibernate-core/src/main/java/org/hibernate/loader/entity/AbstractEntityLoader.java:
```diff
@@ -61,6 +61,11 @@ public Object load(Serializable id, Object optionalObject, SharedSessionContract
+	public Object load(Object id, SharedSessionContractImplementor session, LockOptions lockOptions) {
+		return load( session, id, null, null, lockOptions, null );
+	}
+
```
This is a replacement for `loadByUniqueKey()`.

gavinking commented on hibernate-core/src/main/java/org/hibernate/loader/entity/EntityLoader.java:
```diff
@@ -140,14 +139,6 @@ public EntityLoader(
-	public Object loadByUniqueKey(SharedSessionContractImplementor session, Object key, Boolean readOnly) {
-		return load( session, key, null, null, LockOptions.NONE, readOnly );
-	}
-
```
These really weren't doing a lot of work, and HR can't override them.

Sanne replied:
I'll change this to deprecate them, suggesting to use `load` instead. Ok?

gavinking replied:
Totally fine by me :)

gavinking commented on hibernate-core/src/main/java/org/hibernate/loader/entity/UniqueEntityLoader.java:
```diff
@@ -60,4 +69,11 @@ default Object load(
+			SharedSessionContractImplementor session,
+			LockOptions lockOptions) {
+		throw new UnsupportedOperationException();
+	}
```
This is a replacement for `loadByUniqueKey()`. Unimplemented by default. But in future we could extend unique key loading to other `UniqueEntityLoader`s.

gavinking commented on hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:
```diff
@@ -255,7 +255,7 @@
 	private final EntityLoaderLazyCollection loaders = new EntityLoaderLazyCollection();
 
-	private volatile Map<String,EntityLoader> uniqueKeyLoaders;
+	private volatile Map<String,UniqueEntityLoader> uniqueKeyLoaders;
```
This change is necessary to allow HR to provide its own reactive loaders.

DavideD commented:
Shouldn't this be a GitHub draft?

gavinking commented:
Yes I meant to do that but I forgot.

gavinking commented:
So here's the thing: the overloaded `load()` operations of `UniqueEntityLoader` are really a bit of a mess at this point. So, should I take this as an opportunity to clean it up, or do I just add that one new `default` method and leave everything else alone? (And clean it up for H6.)

WDYT, @Sanne, @sebersole?

Sanne commented:
> So here's the thing: the overloaded load() operations of UniqueEntityLoader are really a bit of a mess at this point. So, should I take this as an opportunity to clean it up, or do I just add that one new default method and leave everything else alone? (And clean it up for H6.)

I guess it depends what you plan?  We can of course trust you and have no problem with small refactorings, but I'd still prefer us being highly conservative with 5.x in case you have semantic changes in mind.

N.B. I'll be mostly offline this week as I'm moving home - but I'm sure you all know what you're doing

sebersole commented:
This is already refactored in 6, so no worry about compatibility - its already something we planned on being not compatible.

However, these are public APIs and probably better to leave alone for 5.x unless there is a clear reason.

gavinking commented:
OK, fine. So then are the changes I've proposed here OK?

I added a `default` method to `UniqueEntityLoader` and removed a couple of methods from `EntityLoader`. Does anyone object to those changes?

Sanne commented:
I'm a bit puzzled to see that @gavinking 's changes are conflicting with patches I wrote ~5 months ago. @gavinking may I suggest to pull changes sometimes :)

gavinking commented:
Uh..... wha?

gavinking commented:
According to GH this branch is only 16 commits behind master. (I sent it two weeks ago.)

Sanne commented:
tracking as https://hibernate.atlassian.net/browse/HHH-14474

Sanne commented:
> According to GH this branch is only 16 commits behind master. (I sent it two weeks ago.)

I got confused as I had made some changes recently as well - still there were conflicts, looks like with other improvements of yours that have been merged in the meantime. FWIW this will look a little different in branch 5.4.

Sanne commented:
This was merged now.

gavinking commented:
Thanks so much @Sanne, appreciated!!

Sanne commented:
thanks @gavinking :)  Could you check your HR branch using a fresh build from ORM branch `5.4` ? 

I had to make some little changes here - should be equivalent to your original patch, but I'd rather verify this before we tag.

gavinking commented:
On it.

gavinking commented on hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:
```diff
@@ -2469,17 +2469,17 @@ private void internalInitSubclassPropertyAliasesMap(String path, Iterator proper
 	}
 
-	private EntityLoader getAppropriateUniqueKeyLoader(String propertyName, SharedSessionContractImplementor session) {
+	protected UniqueEntityLoader getAppropriateUniqueKeyLoader(String propertyName, SharedSessionContractImplementor session) {
```
@Sanne I think we might have lost this change:

`getAppropriateUniqueKeyLoader()` needs to be made visible.

Sanne replied:
Ok, I wondered about that. Isn't access to override `createUniqueKeyLoader` good enough for HR ?

Sanne replied:
Ok, I wondered about that. Isn't access to override `createUniqueKeyLoader` good enough for HR ?

gavinking replied:
No i need it.

Sanne replied:
Ok, done. Thanks!
